app/components/slack-context-menu.js

```diff
@@ -0,0 +1,165 @@
+		target.setSelectionRange(selectionStart,selectionStart); // place cursor at original selectionStart
+	}
+
+	function buildTextareaMenu(menu, target) {
+		var selected = isSelected(target);
+
+		// Select all
+		menu.append(new gui.MenuItem({
+			label: 'Select All',
+			click: function () {
+				target.select();
+			}
+		}));
+		// TODO: Undo
+		// TODO: Redo
+		// Cut
+		menu.append(new gui.MenuItem({
+			label: 'Cut',
+			enabled: selected,
+			click: function () {
+				clipboard.set(removeSelectedValue(target));
+			}
+		}));
+		// Copy
+		menu.append(new gui.MenuItem({
+			label: 'Copy',
+			enabled: selected,
+			click: function () {
+				clipboard.set(getSelectedValue(target));
+			}
+		}));
+		// Paste
+		menu.append(new gui.MenuItem({
+			label: 'Paste',
+			enabled: clipboard.get() != undefined && clipboard.get().trim() != '',
+			click: function () {
+				pasteValue(target, clipboard.get());
+			}
+		}));
+		// Delete
+		menu.append(new gui.MenuItem({
+			label: 'Delete',
+			enabled: selected,
+			click: function () {
+				removeSelectedValue(target);
+			}
+		}));
+		// Search Google
+		if (selected) {
+			var selectedValue = getSelectedValue(target);
+			var labelValue = selectedValue;
+			if (labelValue.length > 10) {
+				labelValue = labelValue.substring(0,10) + "...";
+			}
+			menu.append(new gui.MenuItem({
+				label: 'Search Google for "' + labelValue + '"',
+				click: function() {
+					gui.Shell.openExternal( GOOGLE_SEARCH_URL + encodeURIComponent(selectedValue) );
+				}
+			}));
+		}
+	}
+
+	function buildImageMenu(menu, target) {
+		menu.append(new gui.MenuItem({
+			label: 'Open Image in Browser',
+			click: function() {
+				if (target.attributes['src']) {
+					gui.Shell.openExternal( target.attributes['src'].value );
+				}
+			}
+		}));
+	}
+
+	function buildLinkMenu(menu, target) {
+		if (target.attributes['href'] && target.attributes['href'].value) {
+			var hrefValue = target.attributes['href'].value;
+			if (hrefValue.startsWith("http") || hrefValue.startsWith("https")) {
+				menu.append(new gui.MenuItem({
+					label: 'Open Link in Browser',
+					click: function () {
+						if (target.attributes['href']) {
+							gui.Shell.openExternal(hrefValue);
+						}
+					}
+				}));
+			}
+		}
+	}
+
+	function buildGenericMenu(window, menu) {
+		var selection = window.getSelection();
+		var selected = selection && selection.toString() && selection.toString().trim().length > 0;
+
+		// Search Google
+		if (selected) {
+			var selectedValue = selection.toString();
+			var labelValue = selectedValue;
+			if (labelValue.length > 10) {
+				labelValue = labelValue.substring(0,10) + "...";
+			}
+			menu.append(new gui.MenuItem({
+				label: 'Search Google for "' + labelValue + '"',
+				click: function() {
+					gui.Shell.openExternal( GOOGLE_SEARCH_URL + encodeURIComponent(selectedValue) );
+				}
+			}));
```

Comment:
The code here seems to be copy/pasted from above. Is there any way to make a factory for this (e.g. `createSearchMenuItem(val)